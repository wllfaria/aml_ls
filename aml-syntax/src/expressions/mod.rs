use aml_core::Location;
use aml_token::{Operator, TokenKind, Tokens};

use crate::ast::{ArrayIndex, Binary, Call, ErrorExpr, Expr, List, Map, PrimitiveExpr, Unary};

#[cfg(test)]
mod tests;

#[cfg(test)]
pub(crate) mod snapshots;

pub mod precedences {
    pub const INITIAL: u8 = 0;
    pub const CONDITIONAL: u8 = 2;
    pub const EQUALITY: u8 = 3;
    pub const LOGICAL: u8 = 4;
    pub const SUM: u8 = 5;
    pub const PRODUCT: u8 = 6;
    pub const PREFIX: u8 = 8;
    pub const CALL: u8 = 10;
    pub const SUBCRIPT: u8 = 11;
}

fn get_precedence(op: Operator) -> u8 {
    match op {
        Operator::Dot | Operator::LBracket => precedences::SUBCRIPT,
        Operator::LParen => precedences::CALL,
        Operator::Mul | Operator::Div | Operator::Mod => precedences::PRODUCT,
        Operator::Plus | Operator::Minus => precedences::SUM,
        Operator::GreaterThan
        | Operator::GreaterThanOrEqual
        | Operator::LessThan
        | Operator::LessThanOrEqual => precedences::LOGICAL,
        Operator::EqualEqual | Operator::NotEqual => precedences::EQUALITY,
        Operator::Or | Operator::And | Operator::Either => precedences::CONDITIONAL,

        _ => precedences::INITIAL,
    }
}

pub fn parse_expression(tokens: &mut Tokens) -> Expr {
    parse_expression_inner(tokens, precedences::INITIAL)
}

fn parse_expression_inner(tokens: &mut Tokens, precedence: u8) -> Expr {
    let next = tokens.next_no_indent();
    let location = next.location();

    let mut lhs = match next.kind() {
        TokenKind::Operator(Operator::LBracket) => parse_collection(tokens, location),
        TokenKind::Operator(Operator::LCurly) => parse_map(tokens, location),
        TokenKind::Operator(Operator::LParen) => {
            let lhs = parse_expression_inner(tokens, precedences::INITIAL);
            assert!(matches!(
                tokens.next_no_indent().kind(),
                TokenKind::Operator(Operator::RParen)
            ));
            lhs
        }
        TokenKind::Operator(op @ Operator::Minus) => parse_unary_expression(tokens, op, location),
        TokenKind::Operator(op @ Operator::Not) => parse_unary_expression(tokens, op, location),
        TokenKind::String(location) => Expr::String(location),
        TokenKind::Primitive(primitive) => Expr::Primitive(PrimitiveExpr {
            value: primitive,
            location,
        }),
        TokenKind::Identifier(location) => Expr::Ident(location),

        // all of these are invalid
        TokenKind::Operator(Operator::RBracket)
        | TokenKind::Operator(Operator::RParen)
        | TokenKind::Operator(Operator::Association)
        | TokenKind::Operator(Operator::RCurly)
        | TokenKind::Operator(Operator::Mul)
        | TokenKind::Operator(Operator::Div)
        | TokenKind::Operator(Operator::Mod)
        | TokenKind::Operator(Operator::PlusEqual)
        | TokenKind::Operator(Operator::MinusEqual)
        | TokenKind::Operator(Operator::MulEqual)
        | TokenKind::Operator(Operator::DivEqual)
        | TokenKind::Operator(Operator::ModEqual)
        | TokenKind::Operator(Operator::Colon)
        | TokenKind::Operator(Operator::Comma)
        | TokenKind::Operator(Operator::Plus)
        | TokenKind::Operator(Operator::Dot)
        | TokenKind::Operator(Operator::GreaterThan)
        | TokenKind::Operator(Operator::GreaterThanOrEqual)
        | TokenKind::Operator(Operator::LessThan)
        | TokenKind::Operator(Operator::LessThanOrEqual)
        | TokenKind::Operator(Operator::EqualEqual)
        | TokenKind::Operator(Operator::NotEqual)
        | TokenKind::Operator(Operator::Or)
        | TokenKind::Operator(Operator::And)
        | TokenKind::Operator(Operator::Either)
        | TokenKind::Equal
        | TokenKind::For
        | TokenKind::In
        | TokenKind::If
        | TokenKind::Else
        | TokenKind::Switch
        | TokenKind::Case
        | TokenKind::Default
        | TokenKind::With
        | TokenKind::As
        | TokenKind::Component
        | TokenKind::ComponentSlot
        | TokenKind::Decl
        | TokenKind::Local
        | TokenKind::Global
        | TokenKind::Eof
        | TokenKind::Error(_)
        | TokenKind::Element(_)
        | TokenKind::Container(_)
        | TokenKind::Newline => {
            return Expr::Error(ErrorExpr {
                location,
                token: next.kind(),
            });
        }

        TokenKind::Indent(_) => unreachable!(),
    };

    loop {
        let TokenKind::Operator(op) = tokens.peek_skip_indent().kind() else {
            return lhs;
        };

        match op {
            Operator::Association
            | Operator::PlusEqual
            | Operator::MinusEqual
            | Operator::MulEqual
            | Operator::DivEqual
            | Operator::ModEqual => {
                let token = tokens.next_no_indent();
                return Expr::Error(ErrorExpr {
                    location: token.location(),
                    token: token.kind(),
                });
            }
            _ => {}
        }

        let op_precedence = get_precedence(op);

        if precedence >= op_precedence {
            break;
        }

        tokens.consume();

        match op {
            Operator::LParen => {
                lhs = parse_function(tokens, lhs, location);
                continue;
            }
            Operator::LBracket => {
                let index = parse_expression_inner(tokens, precedences::INITIAL);
                let next_token = tokens.next_no_indent();

                assert!(matches!(
                    next_token.kind(),
                    TokenKind::Operator(Operator::RBracket)
                ));

                lhs = Expr::ArrayIndex(ArrayIndex {
                    lhs: Box::new(lhs),
                    location: location.merge(next_token.location()),
                    index: Box::new(index),
                });

                continue;
            }
            _ => {}
        }

        let rhs = parse_expression_inner(tokens, op_precedence);
        let location = location.merge(rhs.location());
        lhs = Expr::Binary(Binary {
            op,
            lhs: Box::new(lhs),
            rhs: Box::new(rhs),
            location,
        });
    }

    lhs
}

fn parse_unary_expression(tokens: &mut Tokens, operator: Operator, location: Location) -> Expr {
    let expr = parse_expression_inner(tokens, precedences::PREFIX);
    let location = location.merge(expr.location());

    Expr::Unary(Unary {
        op: operator,
        expr: Box::new(expr),
        location,
    })
}

fn parse_collection(tokens: &mut Tokens, start_location: Location) -> Expr {
    let mut items = vec![];

    let end_location = loop {
        let next_token = tokens.peek_skip_indent();
        match next_token.kind() {
            TokenKind::Eof => {
                items.push(Expr::Error(ErrorExpr {
                    location: next_token.location(),
                    token: next_token.kind(),
                }));
                break next_token.location();
            }
            TokenKind::Newline => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::RBracket) => {
                tokens.consume();
                break next_token.location();
            }
            TokenKind::Operator(Operator::Comma) => {
                tokens.consume();
                continue;
            }
            _ => {}
        }
        items.push(parse_expression_inner(tokens, precedences::INITIAL));
    };

    let location = start_location.merge(end_location);
    Expr::List(List { items, location })
}

fn parse_map(tokens: &mut Tokens, start_location: Location) -> Expr {
    let mut items = vec![];

    let end_location = loop {
        let next_token = tokens.peek_skip_indent();
        match next_token.kind() {
            TokenKind::Newline => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::Comma) => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::RCurly) => {
                tokens.consume();
                break next_token.location();
            }
            _ => {}
        }

        let key = parse_expression_inner(tokens, precedences::INITIAL);

        match tokens.peek_skip_indent().kind() {
            TokenKind::Operator(Operator::Colon) => tokens.consume(),
            _ => {
                items.push((
                    key,
                    Expr::Error(ErrorExpr {
                        location: next_token.location(),
                        token: next_token.kind(),
                    }),
                ));
                break next_token.location();
            }
        }
        let value = parse_expression_inner(tokens, precedences::INITIAL);
        items.push((key, value));
    };

    let location = start_location.merge(end_location);

    Expr::Map(Map { items, location })
}

fn parse_function(tokens: &mut Tokens, lhs: Expr, start_location: Location) -> Expr {
    let mut args = vec![];

    let end_location = loop {
        let next_token = tokens.peek_skip_indent();
        match next_token.kind() {
            TokenKind::Operator(Operator::Comma) => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::RParen) => {
                tokens.consume();
                break next_token.location();
            }
            TokenKind::Eof => {
                args.push(Expr::Error(ErrorExpr {
                    location: next_token.location(),
                    token: next_token.kind(),
                }));
                break next_token.location();
            }
            _ => {}
        }
        args.push(parse_expression_inner(tokens, precedences::INITIAL));
    };

    let location = start_location.merge(end_location);

    Expr::Call(Call {
        fun: Box::new(lhs),
        args,
        location,
    })
}
