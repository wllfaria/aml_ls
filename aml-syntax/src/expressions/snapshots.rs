use aml_core::Location;
use aml_token::{Operator, Primitive, TokenKind};
use serde::Serialize;

use crate::ast::*;
use crate::parser::snapshots::ToSnapshot;

impl<'ast> ToSnapshot<'ast> for Expr {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        match self {
            Expr::Ident(location) => SnapshotExpr::Ident(SnapshotIdentifier {
                location,
                value: &content[location.to_range()],
            }),
            Expr::String(location) => SnapshotExpr::String(SnapshotString {
                location,
                value: &content[location.to_range()],
            }),
            Expr::Unary(unary) => unary.into_snapshot(content),
            Expr::Binary(binary) => binary.into_snapshot(content),
            Expr::Call(call) => call.into_snapshot(content),
            Expr::Primitive(primitive) => primitive.into_snapshot(content),
            Expr::ArrayIndex(array_index) => array_index.into_snapshot(content),
            Expr::List(list) => list.into_snapshot(content),
            Expr::Map(map) => map.into_snapshot(content),
            Expr::Error(error) => error.into_snapshot(content),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotUnary<'ast> {
    pub original: &'ast str,
    pub op: Operator,
    pub expr: Box<SnapshotExpr<'ast>>,
    pub location: Location,
}

impl<'ast> ToSnapshot<'ast> for Unary {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Unary(SnapshotUnary {
            op: self.op,
            expr: Box::new(self.expr.into_snapshot(content)),
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotBinary<'ast> {
    pub lhs: Box<SnapshotExpr<'ast>>,
    pub rhs: Box<SnapshotExpr<'ast>>,
    pub op: Operator,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for Binary {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Binary(SnapshotBinary {
            lhs: Box::new(self.lhs.into_snapshot(content)),
            rhs: Box::new(self.rhs.into_snapshot(content)),
            op: self.op,
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotIdentifier<'ast> {
    pub value: &'ast str,
    pub location: Location,
}

impl<'ast> ToSnapshot<'ast> for SnapshotIdentifier<'ast> {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Ident(SnapshotIdentifier {
            location: self.location,
            value: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotString<'ast> {
    pub value: &'ast str,
    pub location: Location,
}

impl<'ast> ToSnapshot<'ast> for SnapshotString<'ast> {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::String(SnapshotString {
            location: self.location,
            value: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotCall<'ast> {
    pub fun: Box<SnapshotExpr<'ast>>,
    pub args: Vec<SnapshotExpr<'ast>>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for Call {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Call(SnapshotCall {
            fun: Box::new(self.fun.into_snapshot(content)),
            args: self
                .args
                .into_iter()
                .map(|e| e.into_snapshot(content))
                .collect(),
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotPrimitive<'ast> {
    pub value: Primitive,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for PrimitiveExpr {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Primitive(SnapshotPrimitive {
            value: self.value,
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotArrayIndex<'ast> {
    pub lhs: Box<SnapshotExpr<'ast>>,
    pub index: Box<SnapshotExpr<'ast>>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for ArrayIndex {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::ArrayIndex(SnapshotArrayIndex {
            location: self.location,
            original: &content[self.location.to_range()],
            lhs: Box::new(self.lhs.into_snapshot(content)),
            index: Box::new(self.index.into_snapshot(content)),
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotList<'ast> {
    pub items: Vec<SnapshotExpr<'ast>>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for List {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::List(SnapshotList {
            location: self.location,
            original: &content[self.location.to_range()],
            items: self
                .items
                .into_iter()
                .map(|e| e.into_snapshot(content))
                .collect(),
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotMap<'ast> {
    pub items: Vec<(SnapshotExpr<'ast>, SnapshotExpr<'ast>)>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for Map {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Map(SnapshotMap {
            location: self.location,
            original: &content[self.location.to_range()],
            items: self
                .items
                .into_iter()
                .map(|(l, r)| (l.into_snapshot(content), r.into_snapshot(content)))
                .collect(),
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotErrorExpr<'ast> {
    pub token: TokenKind,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for ErrorExpr {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Error(SnapshotErrorExpr {
            token: self.token,
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub enum SnapshotExpr<'ast> {
    Unary(SnapshotUnary<'ast>),
    Binary(SnapshotBinary<'ast>),
    Ident(SnapshotIdentifier<'ast>),
    String(SnapshotString<'ast>),
    Call(SnapshotCall<'ast>),
    Primitive(SnapshotPrimitive<'ast>),
    ArrayIndex(SnapshotArrayIndex<'ast>),
    List(SnapshotList<'ast>),
    Map(SnapshotMap<'ast>),
    Error(SnapshotErrorExpr<'ast>),
}
