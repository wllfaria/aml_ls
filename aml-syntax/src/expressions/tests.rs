use aml_token::{Lexer, Tokens};

use super::snapshots::SnapshotExpr;
use crate::expressions::precedences;
use crate::parser::snapshots::ToSnapshot;

fn parse(input: &str) -> SnapshotExpr<'_> {
    let tokens = Lexer::new(input).collect();
    let mut tokens = Tokens::new(tokens, input.len());
    let expression = super::parse_expression_inner(&mut tokens, precedences::INITIAL);
    expression.into_snapshot(input)
}

#[test]
fn test_simple_add() {
    let input = "1 + 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_simple_sub() {
    let input = "1 - 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_simple_mul() {
    let input = "5 + 1 * 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_simple_div() {
    let input = "5 - 1 / 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_expr_grouping_precedence() {
    let input = "(5 + 1) * 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_function_call() {
    let input = "fun(1, a + 2 * 3, 3)";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_function_no_args() {
    let input = "f()";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_array_index() {
    let input = "array[0][1]";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_map_lookup() {
    let input = "map['key']";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_dot_lookup() {
    let input = "a.b.c";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_modulo() {
    let input = "5 + 1 % 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_simple_list() {
    let input = "[1, 2, a, 4]";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_nested_list() {
    let input = "[1, [2, 3, [4, 5]]]";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_simple_equality() {
    let input = "1 == 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_map_declaration() {
    let input = "{a: 1, b: c}";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_and() {
    let input = "1 == 2 && 3 == 4";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_not() {
    let input = "1 != 2 && 3 != 4";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_either() {
    let input = "a ? b ? c";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_unclosed_parenthesis() {
    let input = "func(1, 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_unclosed_bracket() {
    let input = "[1, 2, 3";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_unclosed_curly_brace() {
    let input = "{a: 1, b: 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_missing_comma_in_list() {
    let input = "[1 + 12 * a b c - 13]";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_missing_comma_in_function_args() {
    let input = "foo(1 2 3)";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_missing_operand() {
    let input = "1 + ";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_missing_left_operand() {
    let input = "* 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_mixed_brackets() {
    let input = "(1 + [2)";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_incomplete_binary_expression() {
    let input = "a +";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_double_operators() {
    let input = "1 + + 2";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_empty_map_entry() {
    let input = "{: 1, b:}";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_invalid_map_key() {
    let input = "{1+2: 3}";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_incomplete_dot_lookup() {
    let input = "obj.";
    insta::assert_yaml_snapshot!(parse(input));
}

// TODO(wiru): this also hits an assertion error, which it shouldn't
// #[test]
// fn test_empty_array_index() {
//     let input = "array[]";
//     insta::assert_yaml_snapshot!(parse(input));
// }

#[test]
fn test_invalid_either() {
    let input = "a ? ";
    insta::assert_yaml_snapshot!(parse(input));
}

#[test]
fn test_consecutive_either() {
    let input = "a ? ? c";
    insta::assert_yaml_snapshot!(parse(input));
}
