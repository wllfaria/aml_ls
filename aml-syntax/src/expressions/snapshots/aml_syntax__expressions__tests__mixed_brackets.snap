---
source: aml-syntax/src/expressions/tests.rs
expression: parse(input)
---
Binary:
  lhs:
    Primitive:
      value:
        Int: 1
      location:
        start_byte: 1
        end_byte: 2
      original: "1"
  rhs:
    List:
      items:
        - Primitive:
            value:
              Int: 2
            location:
              start_byte: 6
              end_byte: 7
            original: "2"
        - Error:
            token:
              Operator: RParen
            location:
              start_byte: 7
              end_byte: 8
            original: )
        - Error:
            token: Eof
            location:
              start_byte: 7
              end_byte: 8
            original: )
      location:
        start_byte: 5
        end_byte: 8
      original: "[2)"
  op: Plus
  location:
    start_byte: 1
    end_byte: 8
  original: "1 + [2)"
